### DISCUSSION TITLE
`-output-complete-exe` is unable to produce C source?

### DISCUSSION THREAD

**Discussion by rgrinberg** [ocaml/ocaml#12661] [Q&A]
This session produces a C file:
```
$ touch foo.ml
$ ocamlc -o foo.c foo.ml -output-complete-obj
$ ls foo.c
foo.c
```

While this one does not:
```
$ touch foo.ml
$ ocamlc -o foo.c foo.ml -output-complete-exe
$ ls foo.c
ls: cannot access 'foo.c': No such file or directory
```

Is this difference of behavior between `-output-complete-obj` and `output-complete-exe` intended?

**Comment by lthls** [comment]
The only command that is expected to produce a C file is `-output-complete-obj -o foo.c`. In all other cases, any C files created during compilation are removed afterwards.

But you stumbled upon a bug: with `-output-complete-exe -o foo.c`, the compiler tries to produce an executable called `foo.c`, but the intermediate C file produced during compilation is also called `foo.c`, so depending on your C compiler this could either cause an error (with my version of GCC) or silently overwrite the input file (I guess that's what happens in your case). In the latter case, the OCaml linker will then remove its intermediary file, which happens to also be the final executable.
I guess we should throw an error early when the output name for `-output-complete-exe` ends in `.c` (or a C object extension), since that's likely a misuse of the feature.

**Reply by rgrinberg** [reply]
> The only command that is expected to produce a C file is -output-complete-obj -o foo.c. In all other cases, any C files created during compilation are removed afterwards.

Shouldn't `-output-obj` also be allowed to produce C files?

> I guess we should throw an error early when the output name for -output-complete-exe ends in .c (or a C object extension), since that's likely a misuse of the feature.

Why is it a misuse? Allowing the user to compile or post process the C file could be a valid use case.

**Comment by lthls** [comment]
> Shouldn't `-output-obj` also be allowed to produce C files?

Sorry, yes, `-output-obj` behaves as `-output-complete-obj` except for what goes into the C file.

> Why is it a misuse? Allowing the user to compile or post process the C file could be a valid use case.

But why not use `-output-complete-obj` then ? The C file produced is almost the same.

The main difference between the `-obj` and `-exe` versions is that `-obj` provides you something that you can integrate into something else (either as a source or object file), while `-exe` links everything for you into an executable. If you want the C file produced by `-output-complete-exe`, I think it would make more sense to add a flag (say, `-with-c-main`) such that `-output-complete-obj -with-c-main` produces the same C file as `output-complete-exe`. But again, I don't understand why you need the extra `main` function if you're going to patch the C file anyway.

**Reply by rgrinberg** [reply]
Okay, that makes sense.

**Comment by rgrinberg** [comment]
Closing this as the answers from @lthls are satisfactory. The bug that I've stumbled on doesn't seem very important, so I won't bother reporting it.
